titus-server-master/src/main/java/com/netflix/titus/master/mesos/kubeapiserver/DefaultKubeJobManagementReconciler.java

```diff
@@ -0,0 +1,286 @@
+    }
+
+    /**
+     * Transition orphaned tasks to Finished that don't exist in Kubernetes.
+     */
+    private void transitionOrphanedTasks(Map<String, Task> currentTasks, Set<String> currentPodNames, Map<String, V1Node> nodes) {
+        List<Task> tasksNotInApiServer = currentTasks.values().stream()
+                .filter(t -> shouldTaskBeInApiServer(t) && !currentPodNames.contains(t.getId()))
+                .collect(Collectors.toList());
+
+        Map<OrphanedKind, List<Task>> orphanedTasksByKind = new HashMap<>();
+        for (Task task : tasksNotInApiServer) {
+            if (task.getStatus().getState().equals(TaskState.KillInitiated)) {
+                orphanedTasksByKind.computeIfAbsent(OrphanedKind.KILL_INITIATED, s -> new ArrayList<>()).add(task);
+            } else {
+                if (findNode(task, nodes).isPresent()) {
+                    orphanedTasksByKind.computeIfAbsent(OrphanedKind.UNKNOWN, s -> new ArrayList<>()).add(task);
+                } else {
+                    orphanedTasksByKind.computeIfAbsent(OrphanedKind.NODE_TERMINATED, s -> new ArrayList<>()).add(task);
+                }
+            }
+        }
+
+        orphanedTasksByKind.forEach((kind, tasks) -> {
+            logger.info("Attempting to transition {} orphaned tasks to finished ({}): {}", tasks.size(), kind, tasks);
+            orphanedTaskGauges.get(kind).set(tasks.size());
+
+            for (Task task : tasks) {
+
+                String reasonCode;
+                String reasonMessage;
+                switch (kind) {
+                    case KILL_INITIATED:
+                        reasonCode = TaskStatus.REASON_TASK_KILLED;
+                        reasonMessage = "Task killed";
+                        break;
+                    case NODE_TERMINATED:
+                        reasonCode = TaskStatus.REASON_TASK_LOST;
+                        reasonMessage = "Agent terminated";
+                        break;
+                    case UNKNOWN:
+                    default:
+                        reasonCode = TaskStatus.REASON_TASK_LOST;
+                        reasonMessage = "Task lost between control plane and machine";
+                        break;
+                }
+
+                publishEvent(task,
+                        TaskStatus.newBuilder()
+                                .withState(TaskState.Finished)
+                                .withReasonCode(reasonCode)
+                                .withReasonMessage(reasonMessage)
+                                .withTimestamp(clock.wallTime())
+                                .build()
+                );
+            }
+            logger.info("Finished orphaned task transitions to finished ({})", kind);
+        });
+    }
+
+    private Optional<V1Node> findNode(Task task, Map<String, V1Node> nodes) {
+        String instanceId = task.getTaskContext().get(TaskAttributes.TASK_ATTRIBUTES_AGENT_INSTANCE_ID);
```

Comment:
Instead of doing this, I suggest looking up the pod object and then using the pods nodeName so nodeName can change.

Reply:
Is it something that may happen today, or the goal is to make the code more flexible.
If instance id in the task context is different from node instance id, it is a major consistency issue.

Reply:
It is to make it more flexible as the nodeName in the future may not match the instance id.

Reply:
Adding a new task attribute: 
```java
public static final String TASK_ATTRIBUTES_KUBE_NODE_NAME = "kube.nodeName"
```